web-api/src/storage/elastic/client.rs

```diff
@@ -0,0 +1,315 @@
+            if let Some(value) = value {
+                query_mut.append_pair(key, value);
+            } else {
+                query_mut.append_key_only(key);
+            }
+        }
+        drop(query_mut);
+
+        url
+    }
+
+    pub(super) async fn bulk_request<D>(
+        &self,
+        requests: impl IntoIterator<Item = Result<impl Serialize, serde_json::Error>>,
+    ) -> Result<BulkResponse<D>, Error>
+    where
+        D: DeserializeOwned,
+    {
+        let url = self.create_resource_path(["_bulk"], [("refresh", None)]);
+
+        let mut headers = HeaderMap::new();
+        headers.insert(
+            CONTENT_TYPE,
+            HeaderValue::from_static("application/x-ndjson"),
+        );
+
+        let body = serialize_to_ndjson(requests)?;
+
+        self.query_with_bytes::<_, BulkResponse<D>>(url, Some((headers, body)))
+            .await?
+            .ok_or_else(|| Error::EndpointNotFound("_bulk"))
+    }
+
+    pub(super) async fn query_with_bytes<B, T>(
+        &self,
+        url: Url,
+        post_data: Option<(HeaderMap<HeaderValue>, B)>,
+    ) -> Result<Option<T>, Error>
+    where
+        B: Into<Body>,
+        T: DeserializeOwned,
+    {
+        let request_builder = if let Some((headers, body)) = post_data {
+            self.client.post(url).headers(headers).body(body)
+        } else {
+            self.client.get(url)
+        };
+
+        let response = request_builder
+            .basic_auth(
+                &self.config.user,
+                Some(self.config.password.expose_secret()),
+            )
+            .send()
+            .await?;
+
+        let status = response.status();
+        if status == StatusCode::NOT_FOUND {
+            Ok(None)
+        } else if !status.is_success() {
+            let url = response.url().clone();
+            let body = response.bytes().await?;
+            let error = String::from_utf8_lossy(&body).into_owned();
+            Err(Error::Status { status, url, error })
+        } else {
+            Ok(Some(response.json().await?))
+        }
+    }
+
+    pub(super) async fn query_with_json<B, T>(
+        &self,
+        url: Url,
+        body: Option<B>,
+    ) -> Result<Option<T>, Error>
+    where
+        B: Serialize,
+        T: DeserializeOwned,
+    {
+        let post_data = body
+            .map(|json| -> Result<_, Error> {
+                let mut headers = HeaderMap::new();
+                headers.insert(CONTENT_TYPE, HeaderValue::from_static("application/json"));
+                let body = serde_json::to_vec(&json)?;
+                Ok((headers, body))
+            })
+            .transpose()?;
+
+        self.query_with_bytes(url, post_data).await
+    }
+}
+
+#[derive(Debug, Error, displaydoc::Display, From)]
+pub(crate) enum Error {
+    /// Transmitting a request or receiving the response failed: {0}
+    Transport(reqwest::Error),
+    /// Elastic Search failed, status={status}, url={url}, \nbody={error}
```

Comment:
:art: since this will eventually appear in some server logs we should probably avoid newlines here, iirc this was already an issue in some other part of the code.
```suggestion
    /// Elastic Search failed, status={status}, url={url}, body={error}
```

Reply:
I would argue you do want newlines in long error messages, it's supper common in all kind of software including e.g. what gets logged to Linux system logs.

Reply:
:pushpin: outside of the scope of this PR

Reply:
iirc handling newlines wasn't an issue, it was more about the visual presentation of the logs.